staging/src/k8s.io/kubectl/pkg/cmd/auth/whoami_test.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/cli-runtime/pkg/printers"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/rest/fake"
+	cmdtesting "k8s.io/kubectl/pkg/cmd/testing"
+	"k8s.io/kubectl/pkg/scheme"
+)
+
+func TestWhoAmIRun(t *testing.T) {
+	tests := []struct {
+		name      string
+		o         *WhoAmIOptions
+		args      []string
+		allowed   bool
+		serverErr error
+
+		expectedBodyStrings []string
+	}{
+		{
+			name: "success test",
+			o: &WhoAmIOptions{
+				printer:    printers.NewTablePrinter(printers.PrintOptions{AllowMissingKeys: true}),
+				printTable: true,
+			},
+			args: []string{},
+			expectedBodyStrings: []string{
+				`ATTRIBUTE   VALUE`,
+				`Username    test`,
+				`Groups      [devs admins]`,
+				``,
+			},
+		},
+		{
+			name: "JSON test",
+			o: &WhoAmIOptions{
+				printer: printers.NewTypeSetter(scheme.Scheme).ToPrinter(&printers.JSONPrinter{}),
+			},
+			args: []string{},
+			expectedBodyStrings: []string{
+				`{
+    "kind": "SelfSubjectReview",
+    "apiVersion": "authentication.k8s.io/v1alpha1",
+    "metadata": {
+        "creationTimestamp": null
+    },
+    "status": {
+        "userInfo": {
+            "username": "test",
+            "groups": [
+                "devs",
+                "admins"
+            ]
+        }
+    }
+}
+`,
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			var b bytes.Buffer
+
+			test.o.Out = &b
+			test.o.ErrOut = ioutil.Discard
+
+			tf := cmdtesting.NewTestFactory().WithNamespace("test")
+			defer tf.Cleanup()
+
+			ns := scheme.Codecs.WithoutConversion()
+
+			tf.Client = &fake.RESTClient{
```

Comment:
Instead of manually creating this whole client I'd suggest using something like this https://github.com/kubernetes/kubectl/blob/9aa9a535fcc212780fd51bfc768852a8d0321ab2/pkg/cmd/top/top_pod_test.go#L190-L206 where you focus only on the actions and not whole RESTClient setup.

Reply:
Thanks for this advice. The test became much cleaner.